soroban-spec/src/gen/ts.rs

```diff
@@ -0,0 +1,339 @@
+                "".to_owned()
+            } else if outputs.len() == 1 {
+                format!(": Promise<{}>", type_to_ts(&outputs[0]))
+            } else {
+                format!(
+                    ": Promise<[{}]>>",
+                    outputs.iter().map(type_to_ts).join(", ")
+                )
+            };
+            let ts_doc = doc_to_ts_doc(doc);
+
+            // let output_parser = outputs.get(0).map(scVal_to_type).unwrap_or_default();
+            let output = (!outputs.is_empty())
+                .then(|| {
+                    format!(
+                        r#"
+    return scValToJs(response.xdr) as {}
+"#,
+                        type_to_ts(&outputs[0])
+                    )
+                })
+                .unwrap_or_default();
+            let args = (!inputs.is_empty())
+                .then(|| format!("args: [{args}], "))
+                .unwrap_or_default();
+            format!(
+                r#"{ts_doc}export async function {name}({input}){return_type} {{
+    let invokeArgs: InvokeArgs = {{{sign_me}method: '{name}', {args}}};
+    // @ts-ignore Type does exist
+    const response = await invoke(invokeArgs);{output}
+    
+}}
+"#
+            )
+        }
+        Entry::Struct { doc, name, fields } => {
+            let docs = doc_to_ts_doc(doc);
+            let arg_name = name.to_lower_camel_case();
+            let encoded_fields = js_to_xdr_fields(&arg_name, fields);
+            let fields = fields.iter().map(field_to_ts).join("\n  ");
+            let void = type_to_js_xdr(&Type::Void);
+            format!(
+                r#"{docs}export interface {name} {{
+  {fields}
+}}
+
+function {name}ToXDR({arg_name}?: {name}): xdr.ScVal {{
+    if (!{arg_name}) {{
+        return {void};
+    }}
+    let arr = [
+        {encoded_fields}
+        ];
+    return xdr.ScVal.scvMap(arr);
+}}
+"#
+            )
+        }
+
+        Entry::Union { name, doc, cases } => {
+            let doc = doc_to_ts_doc(doc);
+            let arg_name = name.to_lower_camel_case();
+            let encoded_cases = js_to_xdr_union_cases(&arg_name, cases);
+            let cases = cases.iter().map(case_to_ts).join(" | ");
+            let void = type_to_js_xdr(&Type::Void);
+
+            format!(
+                r#"{doc}export type {name} = {cases};
+
+function {name}ToXDR({arg_name}?: {name}): xdr.ScVal {{
+    if (!{arg_name}) {{
+        return {void};
+    }}
+    let res: xdr.ScVal[] = [];
+    switch ({arg_name}.tag) {{
+        {encoded_cases}  
+    }}
+    return xdr.ScVal.scvVec(res);
+}}
+"#
+            )
+        }
+        Entry::Enum { doc, name, cases } => {
+            if name == "Error" {
+                let cases = cases
+                    .iter()
+                    .map(|c| format!("{{error:\"{}\"}}", c.doc))
+                    .join(",\n  ");
+                return format!(
+                    r#"const Errors = [ 
+  {cases}
+]"#
+                );
```

Comment:
What's the `Error` special case for? Contract errors should be separated into a different type `ScSpecErrorEnum`, and shouldn't show up as `ScSpecEnum`.